torch/_torch_docs.py

```diff
@@ -8450,50 +8450,53 @@ def merge_dicts(*dicts):
-          `gesvdj` and `gesvdjBatched` on CUDA 10.1.243 and later, and uses the MAGMA routine `gesdd`
-          on earlier versions of CUDA.
+.. note:: The implementation of SVD on CPU uses the LAPACK routine ``?gesdd`` (a divide-and-conquer
+          algorithm) instead of ``?gesvd`` for speed. Analogously, the SVD on GPU uses the
+          cuSOLVER routines ``gesvdj`` and ``gesvdjBatched`` on CUDA 10.1.243 and later,
+          and uses the MAGMA routine ``gesdd`` on earlier versions of CUDA.
 
-.. note:: The returned matrix `U` will be transposed, i.e. with strides
-          :code:`U.contiguous().transpose(-2, -1).stride()`.
+.. note:: The returned matrix ``U`` will not be contiguous. It will be represented as a
+          column-major matrix (i.e. fortran-contiguous).
 
-.. note:: Gradients computed using `U` and `V` may be unstable if
-          :attr:`input` is not full rank or has non-unique singular values.
+.. note:: Gradients computed using ``U`` and ``V`` may be unstable if :attr:`input` has
+          repeated non-unique singular values, e.g., when it is not full-rank.
+          See also the last note in this list.
 
-.. note:: When :attr:`some` = ``False``, the gradients on :code:`U[..., :, min(m, n):]`
-          and :code:`V[..., :, min(m, n):]` will be ignored in backward as those vectors
+.. note:: When :attr:`some` is ``False``, the gradients on ``U[..., :, min(m, n):]``
+          and ``V[..., :, min(m, n):]`` will be ignored in backward as those vectors
           can be arbitrary bases of the subspaces.
 
-.. note:: The `S` tensor can only be used to compute gradients if :attr:`compute_uv` is True.
-
-.. note:: With the complex-valued input the backward operation works correctly only
-          for gauge invariant loss functions. Please look at `Gauge problem in AD`_ for more details.
+.. note:: The ``S`` tensor can only be used to compute gradients if :attr:`compute_uv` is ``True``.
 
-.. note:: Since `U` and `V` of an SVD is not unique, each vector can be multiplied by
-          an arbitrary phase factor :math:`e^{i \phi}` while the SVD result is still correct.
-          Different platforms, like Numpy, or inputs on different device types, may produce different
-          `U` and `V` tensors.
+.. note:: For complex-valued :attr:`input` the SVD is not unique, as ``U`` and ``V`` may
+          be multiplied by an arbitrary phase factor :math:`e^{i \phi}` on every column.
+          The same happens when :attr:`input` has repeated singular values, where one may multiply
+          the columns of the spanning subspace in ``U`` and ``V`` by a rotation matrix
+          and the resulting vectors will span the same subspace.
+          Different platforms, like Numpy, or inputs on different device types,
+          may produce different ``U`` and ``V`` tensors.
+          In these cases, the backward operation is only well-defined when the cost-function
```

Comment:
For torch.linalg.svd we should clarify what actually happens. When say the gradient is only "well-defined" in certain cases we're not actually telling the reader what torch.svd() or torch.linalg.svd() does. Does it compute the gradient when the gradient is well-defined? What does it compute when the gradient is not well-defined? Does it throw a runtime error?

Reply:
That is a great question. The answer is: I do not know.

I have been trying to undestand this problem better, which lead to the discussion in https://github.com/pytorch/pytorch/issues/47599#issuecomment-800920251
It happens to be a surprisingly tricky problem, and I do not understand it well-enough just yet. I will come back to this once I do.

Reply:
Appreciating what we don't know is a great start.

Maybe we can unblock this PR by:

- leaving this issue open for future resolution
- saying that produced gradients are undefined unless the grad passed to svd has certain properties (this would require some wordsmithing)